gctree/branching_processes.py

```diff
@@ -987,3 +964,230 @@ def f(x):
+    Returns:
+        Log likelihood :math:`\ell(p, q; T, A)` and its gradient :math:`\nabla\ell(p, q; T, A)`
+    """
+    terms = [[lltree(cmcounts, p, q), count] for cmcounts, count in cm_countlist]
+    ls = np.array([term[0][0] for term in terms])
+    grad_ls = np.array([term[0][1] for term in terms])
+    # This can be done ahead of time
+    count_ls = np.array([term[1] for term in terms])
+    if marginal:
+        # we need to find the smallest derivative component for each
+        # coordinate, then subtract off to get positive things to logsumexp
+        grad_l = []
+        for j in range(len((p, q))):
+            i_prime = grad_ls[:, j].argmin()
+            b = (grad_ls[:, j] - grad_ls[i_prime, j]) * count_ls
+            # believe it or not, logsumexp can't handle 0 in b
+            # when np.seterr(underflow='raise') on newer processors:
+            grad_l.append(
+                grad_ls[i_prime, j]
+                + np.exp(
+                    scs.logsumexp((ls - ls[i_prime])[b != 0], b=b[b != 0])
+                    - scs.logsumexp(ls - ls[i_prime], b=count_ls)
+                )
+            )
+        # count_ls shouldn't have any zeros in it...
+        return (-np.log(count_ls.sum()) + scs.logsumexp(ls, b=count_ls)), np.array(
+            grad_l
+        )
+    else:
+        return (ls * count_ls).sum(), np.array(
+            [(grad_ls[:, 0] * count_ls).sum(), (grad_ls[:, 1] * count_ls).sum()]
+        )
+
+
+def fit_branching_process(dag, verbose=True, marginal=True):
+    r"""fit p and q using all trees in the dag. DAG should be abundance_annotated, like that output by phylip_parse.
+    if we get floating point errors, try a few more times
+    (starting params aren't random right now, but they could be in the future?)"""
+    cmcount_dagfuncs = cmcounter_dagfuncs()
+    cmcounters = dag.weight_count(**cmcount_dagfuncs)
+
+    def to_tuple(mset):
+        # lltree checks first item in list for unobserved root unifurcation,
+        # but here it could end up not being first.
+        if (0, 1) in mset:
+            assert mset[(0, 1)] == 1
+            mset = mset - {(0, 1)} + {(1, 1)}
+        return tuple(mset.items())
+
+    cmcountlist = [(to_tuple(mset), mult) for mset, mult in cmcounters.items()]
+
+    max_tries = 10
+    for tries in range(max_tries):
+        try:
+            p, q = _mle_helper(
+                lambda p, q: llforest(cmcountlist, p, q, marginal=marginal)
+            )
+            break
+        except FloatingPointError as e:
+            if tries + 1 < max_tries and verbose:
+                print(
+                    f"floating point error in MLE: {e}. "
+                    f"Attempt {tries + 1} of {max_tries}. "
+                    "Rerunning with new random start."
+                )
+            else:
+                raise
+        else:
+            raise
+    return (p, q)
+
+
+def clade_tree_to_ctree(
```

Comment:
Does this need to be a public function in the API that is shown in docs? If not, suggest renaming `_clade_tree_to_ctree`

Reply:
See reply to last comment

Reply:
(and missing type hint on `clade_tree` argument)